File: 11.py
from collections import defaultdict
from math import prod

def printg(grid, mx, my):
    for y in range(0, mx + 1):
        for x in range(0, my + 1):
           if grid[(x, y)] == 0:
               print('.', end='')
           else:
               print(grid[(x, y)], end='')
        print('')

grid = defaultdict(lambda:-1)

max_x = 0
max_y = 0
with open('11.txt') as f:
    y = 0
    for line in (x.strip() for x in f.readlines()):
        x = 0
        for c in line:
            grid[(x,y)] = int(c)
            x += 1
        y += 1
    max_y = y - 1
    max_x = x - 1

adjmap = [ (-1, 0), (0, -1), (1, 0), (0, 1), (-1, -1), (1, -1), (-1, 1), (1, 1) ]

flashes = 0

for step in range(0, 1000):

    #printg(grid, max_x, max_y)
    #input()

    for y in range(0, max_y + 1):
        for x in range(0, max_x + 1):
            grid[(x,y)] += 1

    flashed = set()
    while True:
        flashed_this_round = 0
        for y in range(0, max_y + 1):
            for x in range(0, max_x + 1):
                coord = (x,y)
                v = grid[coord]

                if coord not in flashed and v > 9:
                    flashed.add(coord)
                    flashed_this_round += 1
                    flashes += 1

                    for a in adjmap:
                        nc = (x+a[0], y+a[1])
                        grid[nc] += 1

        if flashed_this_round == 0:
            break

    for c in flashed:
        grid[c] = 0

    if step == 100:
        print('part1', flashes)

    any_non_zero = False
    for y in range(0, max_y + 1):
        for x in range(0, max_x + 1):
            if grid[(x,y)] != 0:
                any_non_zero = True
        if any_non_zero:
            break

    if not any_non_zero:
        print('part1', step + 1)
        break

File: 11.clj
(load-file "utils.clj")
(load-file "grids.clj")

(def grid (grids/parse-int-grid (utils/read-lines "11.txt")))

(defn find-adj-coord [grid x y]
  (remove
    #(nil? (grid %1))
    (map
      (fn[adj]
        [(+ x (adj 0)) (+ y (adj 1))])
      grids/adjacent-list-all)))

(defn count-flashes [grid flashed]
  (let [will-flash (filter #(and (not (contains? flashed (%1 0))) (> (%1 1) 9)) grid)
        new-flashed (into #{} (map #(%1 0) will-flash))]
    (if (empty? new-flashed)
      [grid flashed]
      (count-flashes
        (reduce
          (fn[a c] 
            (if (contains? flashed c)
              a
              (assoc a c (inc (a c)))))
          grid
          (apply concat (for [f will-flash]
            (find-adj-coord grid ((f 0) 0) ((f 0) 1)))))
        (apply conj flashed new-flashed)))))

(defn run-step [grid]
  (let [inc-grid (reduce (fn [a v] (assoc a (v 0) (inc (v 1)))) {} grid)
        r (count-flashes inc-grid #{})
        next-grid (r 0)
        flashed (r 1)]
    [(into next-grid (for [f flashed] [f 0])) flashed]))

(def g (grid :grid))
(def ng
  (reduce 
    (fn[a v]
      (let [r (run-step (a 0))]
        [(r 0) (+ (a 1) (count (r 1)))]))
    [g 0]
    (range 100)))

;(grids/print-grid (merge grid {:grid (ng 0)}))
(println "part1" (ng 1))
